File: module2/class1/exemplo15.cs
/*
================================================================================
ATIVIDADE PRÁTICA 15 - LINQ PERFORMANCE ANTI-PATTERNS (C#)
================================================================================

OBJETIVO:
- Demonstrar anti-patterns de LINQ que causam performance issues
- Usar CPU profiler para identificar multiple enumeration overhead
- Otimizar usando ToList(), caching, e LINQ eficiente
- Medir impacto de multiple enumerations

PROBLEMA:
- Multiple enumeration de IEnumerable causa re-execução custosa
- Chains complexas de LINQ sem materialização são ineficientes
- CPU Profiler mostrará tempo gasto em repeated enumerations

SOLUÇÃO:
- Materializar results com ToList() quando necessário
- Evitar repeated enumerations

================================================================================
*/

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

class Program {
    static void DemonstrateLinqAntiPatterns() {
        Console.WriteLine("Starting LINQ anti-patterns demonstration...");
        Console.WriteLine("Monitor CPU profiler - should see repeated enumeration overhead");
        
        const int DATA_SIZE = 100000;
        
        // Create test data
        var numbers = Enumerable.Range(1, DATA_SIZE);
        
        var sw = Stopwatch.StartNew();
        
        // PERFORMANCE ISSUE: Multiple enumeration of expensive LINQ chain
        var expensiveQuery = numbers
            .Where(x => x % 2 == 0)              // Filter even numbers
            .Select(x => x * x)                  // Square them
            .Where(x => x > 1000)                // Filter large squares
            .Select(x => Math.Sqrt(x));          // Expensive sqrt operation
        
        // PERFORMANCE ISSUE: Each of these operations re-executes the entire chain
        int count = expensiveQuery.Count();              // Enumeration #1
        double sum = expensiveQuery.Sum();               // Enumeration #2  
        double average = expensiveQuery.Average();       // Enumeration #3
        double max = expensiveQuery.Max();               // Enumeration #4
        double min = expensiveQuery.Min();               // Enumeration #5
        
        // PERFORMANCE ISSUE: More re-enumeration
        var sortedResults = expensiveQuery.OrderBy(x => x).Take(10); // Enumeration #6
        var topResults = sortedResults.ToArray();                    // Enumeration #7
        
        sw.Stop();
        
        Console.WriteLine($"LINQ anti-patterns completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Count: {count}, Sum: {sum:F2}, Avg: {average:F2}");
        Console.WriteLine($"Max: {max:F2}, Min: {min:F2}");
        Console.WriteLine($"Top 10 results: {string.Join(", ", topResults.Take(5).Select(x => x.ToString("F1")))}...");
        Console.WriteLine($"Expensive query enumerated ~7 times (massive overhead!)");
    }
    
    static void Main() {
        Console.WriteLine("Starting LINQ performance demonstration...");
        Console.WriteLine("Task: Processing data with inefficient LINQ patterns");
        Console.WriteLine("Monitor CPU Usage Tool for repeated enumeration overhead");
        Console.WriteLine();
        
        DemonstrateLinqAntiPatterns();
        
        Console.WriteLine();
        Console.WriteLine("=== PROFILING ANALYSIS ===");
        Console.WriteLine("Check CPU profiler for:");
        Console.WriteLine("- Repeated execution of LINQ operations");
        Console.WriteLine("- Multiple enumerations of the same query");
        Console.WriteLine("- Wasted CPU cycles on redundant calculations");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR LINQ EFICIENTE)
================================================================================

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

class Program {
    static void DemonstrateEfficientLinq() {
        Console.WriteLine("Starting efficient LINQ demonstration...");
        Console.WriteLine("Monitor CPU profiler - should see single enumeration");
        
        const int DATA_SIZE = 100000;
        
        // Create test data
        var numbers = Enumerable.Range(1, DATA_SIZE);
        
        var sw = Stopwatch.StartNew();
        
        // CORREÇÃO: Materialize expensive query once with ToList()
        var expensiveQuery = numbers
            .Where(x => x % 2 == 0)              // Filter even numbers
            .Select(x => x * x)                  // Square them  
            .Where(x => x > 1000)                // Filter large squares
            .Select(x => Math.Sqrt(x))           // Expensive sqrt operation
            .ToList();                           // CORREÇÃO: Materialize once!
        
        // CORREÇÃO: Now all operations work on materialized list - no re-enumeration
        int count = expensiveQuery.Count;                    // O(1) - uses Count property
        double sum = expensiveQuery.Sum();                   // Single enumeration
        double average = expensiveQuery.Average();           // Single enumeration
        double max = expensiveQuery.Max();                   // Single enumeration
        double min = expensiveQuery.Min();                   // Single enumeration
        
        // CORREÇÃO: Working with materialized data
        var sortedResults = expensiveQuery.OrderBy(x => x).Take(10).ToArray(); // Single sort + take
        
        sw.Stop();
        
        Console.WriteLine($"Efficient LINQ completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Count: {count}, Sum: {sum:F2}, Avg: {average:F2}");
        Console.WriteLine($"Max: {max:F2}, Min: {min:F2}");
        Console.WriteLine($"Top 10 results: {string.Join(", ", sortedResults.Take(5).Select(x => x.ToString("F1")))}...");
        Console.WriteLine($"Expensive query enumerated only ONCE (optimized!)");
    }
    
    static void DemonstrateStreamingLinq() {
        Console.WriteLine("Starting streaming LINQ for large datasets...");
        
        const int DATA_SIZE = 1000000; // Larger dataset
        
        var sw = Stopwatch.StartNew();
        
        // CORREÇÃO: For very large datasets, use streaming approach
        var result = Enumerable.Range(1, DATA_SIZE)
            .Where(x => x % 1000 == 0)           // Early filtering
            .Select(x => x * 2)                  // Simple transform
            .Take(100)                           // CORREÇÃO: Take early to limit processing
            .ToArray();                          // Materialize only final small result
        
        sw.Stop();
        
        Console.WriteLine($"Streaming LINQ completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Processed {DATA_SIZE:N0} items, returned {result.Length} results");
        Console.WriteLine($"Early filtering and Take() prevented unnecessary processing");
    }
    
    static void Main() {
        Console.WriteLine("Starting optimized LINQ demonstration...");
        Console.WriteLine("Task: Processing data with efficient LINQ patterns");
        Console.WriteLine("Monitor CPU Usage Tool for improved performance");
        Console.WriteLine();
        
        DemonstrateEfficientLinq();
        Console.WriteLine();
        DemonstrateStreamingLinq();
        
        Console.WriteLine();
        Console.WriteLine("=== OPTIMIZATION RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- Single enumeration vs multiple re-enumerations");
        Console.WriteLine("- Materialization with ToList() eliminates redundant work");
        Console.WriteLine("- Early filtering reduces processing overhead");
        Console.WriteLine("- Take() limits unnecessary computations");
        Console.WriteLine("- Dramatically improved performance");
    }
}

================================================================================
*/

File: module2/class1/exemplo15.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 15 - CACHE MISS PERFORMANCE (C++)
================================================================================

OBJETIVO:
- Demonstrar impacto de cache misses em performance
- Usar CPU profiler para identificar cache-unfriendly memory access patterns
- Otimizar melhorando data locality e access patterns
- Comparar row-major vs column-major matrix traversal

PROBLEMA:
- Acesso não-sequencial à memória causa cache misses
- Column-major traversal em row-major data é cache-unfriendly
- CPU profiler mostrará alto cache miss ratio

SOLUÇÃO:
- Usar data layout que favoreça cache locality
- Sequential memory access patterns

================================================================================
*/

#include <iostream>
#include <chrono>
#include <vector>
#include <random>
using namespace std;

void demonstrateCacheMisses() {
    cout << "Starting cache miss demonstration..." << endl;
    cout << "Monitor CPU profiler - should see high cache miss ratio" << endl;
    
    const int MATRIX_SIZE = 2000;
    const int ITERATIONS = 3;
    
    // Allocate large matrix
    vector<vector<int>> matrix(MATRIX_SIZE, vector<int>(MATRIX_SIZE));
    
    // Initialize matrix
    for (int i = 0; i < MATRIX_SIZE; i++) {
        for (int j = 0; j < MATRIX_SIZE; j++) {
            matrix[i][j] = i * j;
        }
    }
    
    auto start = chrono::high_resolution_clock::now();
    
    long long sum = 0;
    
    // PERFORMANCE ISSUE: Column-major traversal of row-major data
    // This creates cache misses because we're jumping between cache lines
    for (int iter = 0; iter < ITERATIONS; iter++) {
        for (int j = 0; j < MATRIX_SIZE; j++) {        // Column first
            for (int i = 0; i < MATRIX_SIZE; i++) {    // Row second - BAD for cache
                sum += matrix[i][j]; // Cache miss - different cache line each access
            }
        }
        
        if (iter == 0) {
            cout << "Completed iteration " << (iter + 1) << "/" << ITERATIONS 
                 << " (cache-unfriendly access pattern)" << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Cache-unfriendly traversal completed in: " << duration.count() << " ms" << endl;
    cout << "Sum: " << sum << " (to prevent optimization)" << endl;
    cout << "Access pattern: Column-major on row-major data (cache misses)" << endl;
}

int main() {
    cout << "Starting cache performance demonstration..." << endl;
    cout << "Task: Matrix traversal with cache-unfriendly access pattern" << endl;
    cout << "Monitor CPU Usage Tool and cache performance counters" << endl << endl;
    
    demonstrateCacheMisses();
    
    cout << endl << "=== PROFILING ANALYSIS ===" << endl;
    cout << "Check CPU profiler for:" << endl;
    cout << "- High cache miss ratio" << endl;
    cout << "- Memory stall cycles" << endl;
    cout << "- Poor memory bandwidth utilization" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR CACHE-FRIENDLY ACCESS)
================================================================================

#include <iostream>
#include <chrono>
#include <vector>
using namespace std;

void demonstrateCacheFriendly() {
    cout << "Starting cache-friendly demonstration..." << endl;
    cout << "Monitor CPU profiler - should see improved cache hit ratio" << endl;
    
    const int MATRIX_SIZE = 2000;
    const int ITERATIONS = 3;
    
    // CORREÇÃO: Use single-dimensional array for better cache locality
    vector<int> matrix(MATRIX_SIZE * MATRIX_SIZE);
    
    // Initialize matrix
    for (int i = 0; i < MATRIX_SIZE; i++) {
        for (int j = 0; j < MATRIX_SIZE; j++) {
            matrix[i * MATRIX_SIZE + j] = i * j;
        }
    }
    
    auto start = chrono::high_resolution_clock::now();
    
    long long sum = 0;
    
    // CORREÇÃO: Row-major traversal matches data layout - cache friendly
    for (int iter = 0; iter < ITERATIONS; iter++) {
        for (int i = 0; i < MATRIX_SIZE; i++) {        // Row first
            for (int j = 0; j < MATRIX_SIZE; j++) {    // Column second - GOOD for cache
                sum += matrix[i * MATRIX_SIZE + j]; // Cache hit - sequential access
            }
        }
        
        if (iter == 0) {
            cout << "Completed iteration " << (iter + 1) << "/" << ITERATIONS 
                 << " (cache-friendly access pattern)" << endl;
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Cache-friendly traversal completed in: " << duration.count() << " ms" << endl;
    cout << "Sum: " << sum << " (to prevent optimization)" << endl;
    cout << "Access pattern: Row-major sequential (cache hits)" << endl;
}

void demonstrateBlockedTraversal() {
    cout << "Starting cache-optimized blocked traversal..." << endl;
    
    const int MATRIX_SIZE = 2000;
    const int BLOCK_SIZE = 64; // Cache line friendly block size
    
    vector<int> matrix(MATRIX_SIZE * MATRIX_SIZE);
    
    // Initialize
    for (int i = 0; i < MATRIX_SIZE; i++) {
        for (int j = 0; j < MATRIX_SIZE; j++) {
            matrix[i * MATRIX_SIZE + j] = i + j;
        }
    }
    
    auto start = chrono::high_resolution_clock::now();
    
    long long sum = 0;
    
    // CORREÇÃO: Blocked traversal for even better cache utilization
    for (int bi = 0; bi < MATRIX_SIZE; bi += BLOCK_SIZE) {
        for (int bj = 0; bj < MATRIX_SIZE; bj += BLOCK_SIZE) {
            // Process block
            for (int i = bi; i < min(bi + BLOCK_SIZE, MATRIX_SIZE); i++) {
                for (int j = bj; j < min(bj + BLOCK_SIZE, MATRIX_SIZE); j++) {
                    sum += matrix[i * MATRIX_SIZE + j];
                }
            }
        }
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Blocked traversal completed in: " << duration.count() << " ms" << endl;
    cout << "Sum: " << sum << endl;
}

int main() {
    cout << "Starting optimized cache performance demonstration..." << endl;
    cout << "Task: Matrix traversal with cache-friendly patterns" << endl;
    cout << "Monitor CPU Usage Tool for improved cache performance" << endl << endl;
    
    demonstrateCacheFriendly();
    cout << endl;
    demonstrateBlockedTraversal();
    
    cout << endl << "=== OPTIMIZATION RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- Sequential memory access improves cache hit ratio" << endl;
    cout << "- Blocked traversal maximizes cache line utilization" << endl;
    cout << "- Better memory bandwidth utilization" << endl;
    cout << "- Significantly faster execution time" << endl;
    
    return 0;
}

================================================================================
*/
